Make replace_java_nondet() work even if remove_returns() hasn't been run

owen-mc-diffblue opened the pull request:
Currently, it finds a function call to a function that should be replaced with a nondet, finds the assignment of the return value on the next line, but then goes and finds a second assignment whose rhs is the lhs of the first assignment, and deletes everything between the function call and the second assignment, replacing it with an assignment of the relevant nondet to the lhs of the second assignment.

I don't understand why it does this, so I've changed it to just use the first assignment. If there is a good reason it does what it currently does then please let me know.

tautschnig commented:
I have no idea what this code is doing, mainly because I don't know what the code generated by the Java front-end looks like. The code in question was originally merged in #691. As the resulting code should look different, there ought to be a regression test accompanying this change to show what has change (and how this is better than before).

smowton commented:
@owen-jones-diffblue looks like there's a very particular family of broken tests, and the failures are real (verification success expected, actually got failure)

owen-mc-diffblue commented:
I understand what is going on now. I hope to have an update to this PR soon.

owen-mc-diffblue commented:
The purpose of this PR has changed, so I have changed the title. The behaviour when `remove_returns` has been run, as is the case in `cbmc`, should not change.

tautschnig commented:
Thank you very much for all  the additional explanations! There are three lines the linter complains about:
```
src/goto-programs/replace_java_nondet.cpp:192:  assert is deprecated, use INVARIANT, PRECONDITION, CHECK_RETURN, etc. instead  [build/deprecated] [4]
src/goto-programs/replace_java_nondet.cpp:215:  Put braces on a separate next line  [whitespace/braces] [5]
src/goto-programs/replace_java_nondet.cpp:219:  assert is deprecated, use INVARIANT, PRECONDITION, CHECK_RETURN, etc. instead  [build/deprecated] [4]
```

owen-mc-diffblue commented:
Thanks for pointing those out @tautschnig, I would have missed them. I recall that the linter and clang-format disagreed about formatting braces for lambdas, but not what the resolution was (if there was one). I've tried making the linter happy.

I should have said, the new version of the PR was done by @JohnDumbell, so all credit for improvements should go to him. I'm only pushing it because it's easier for me to push to my fork and keep the same PR.

tautschnig commented on src/goto-programs/replace_java_nondet.cpp:
```diff
@@ -161,40 +174,64 @@ static goto_programt::targett check_and_replace_target(
+  else
+  {
+    // If not, we need to look at the next line instead.
+    assert(next_instr->is_assign());
```
This will need to become a `DATA_INVARIANT`

tautschnig reviewed (commented):
Yes, the linter and clang-format disagree on lambdas;  I think for this one the resolution is fine as it simply leaves some lines untouched.  Otherwise use clang-format's suggestion and add `// NOLINT(whitespace/braces)`.

owen-mc-diffblue commented:
I've replaced the assertions and slightly clarified the function comment.

tautschnig commented:
Based on github rules this could get merged as soon as CI  passes, but I am wondering whether you want to make sure no none-public tools have regressions? Consequently I'd prefer to leave merging to Diffblue team members.

thk123 reviewed (commented):
You know what this comment is going to say - can we have a test of some description? An end-to-end might be quite difficult if remove-returns is always ran, but perhaps a trivial goto program being fed into this analysis to just exercise the new branch.  

In future could reformating/swapping asserts to invariant changes be in a separate commit so easier to see how small the change is. 

:+1: TG pointer bump passed (diffblue/test-gen#1661)

owen-mc-diffblue commented:
@thk123 It is very hard to write a test in this repository, because the `remove_returns` pass is run before the `replace_java_nondet` in jbmc. The only way I can think to do it is to add a flag to run them in the other order, which would obviously be awful. However, the new configuration will be tested in our private repo, so it will be exercised.

tautschnig commented:
> It is very hard to write a test in this repository, because the `remove_returns` pass is run before the `replace_java_nondet` in jbmc. 

How about doing a unit test instead? (See also https://github.com/diffblue/cbmc/pull/1915#issuecomment-379206111 in case it helps.)

thk123 commented on unit/java_bytecode/java_replace_nondet/replace_nondet.cpp:
```diff
@@ -0,0 +1,109 @@
+TEST_CASE(
+  "Load class with a generated java nondet method call, run remove returns "
+  "both before and after the nondet statements have been removed, check results "
+  "are as expected.", "[!mayfail][core][java_bytecode][replace_nondet]")
```
Why `!mayfail`?

JohnDumbell replied:
The mayfail was an oversight on my part - I was writing it on a branch that still had the old code (and so broke).

thk123 reviewed (commented):
This test would be better if if it also checked that the function call hasn't just been removed but actually replaced with the non det expression. You might find the utilities in `require_goto_statements` and `require_expr` useful for this. Appologies if I've jumped the gun with this review

owen-mc-diffblue commented:
@JohnDumbell has added a unit test, using the Catch framework. CI has passed. @thk123 Please review and merge if you approve.

thk123 commented on unit/java_bytecode/java_replace_nondet/replace_nondet.cpp:
```diff
@@ -0,0 +1,120 @@
+    }
+  }
+
+  REQUIRE((method_removed && replacement_nondet_exists));
```
Not blocking but in future I'd split this into two require statements - then the output will be more useful (it will say which one is not true).

thk123 replied:
Thanks :+1:

Commit:
Make remove_java_nondet work before remove_returns

Previously, remove_java_nondet assumed that remove_returns had been run.
Now it will work even if remove_returns hasn't been run.

Commit:
Add unit test for java_replace_nondet
